src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,1005 @@
+                        behavior.Tip = newTip;
+                        this.pullerBehaviorsByPeerId.Add(peerId, behavior);
+
+                        this.logger.LogTrace("New peer with ID {0} and tip '{1}' was added.", peerId, newTip);
+                    }
+                    else
+                        this.logger.LogTrace("Peer ID {0} was discarded since he doesn't support the requirements.", peerId);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void PeerDisconnected(int peerId)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(peerId), peerId);
+
+            lock (this.peerLock)
+            {
+                this.pullerBehaviorsByPeerId.Remove(peerId);
+            }
+
+            this.ReleaseAndReassignAssignments(peerId);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc/>
+        public void RequestBlocksDownload(List<ChainedHeader> headers)
+        {
+            this.logger.LogTrace("({0}:{1})", nameof(headers.Count), headers.Count);
+
+            lock (this.queueLock)
+            {
+                // Enqueue new download job.
+                int jobId = this.nextJobId++;
+
+                this.downloadJobsQueue.Enqueue(new DownloadJob()
+                {
+                    Headers = new List<ChainedHeader>(headers),
+                    Id = jobId
+                });
+
+                this.logger.LogDebug("{0} blocks were requested from puller. Job ID {1} was created.", headers.Count, jobId);
+                this.processQueuesSignal.Set();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Loop that assigns download jobs to the peers.</summary>
+        private async Task AssignerLoopAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellationSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.processQueuesSignal.WaitAsync(this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+
+                await this.AssignDownloadJobsAsync().ConfigureAwait(false);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Loop that continuously checks if peers failed to deliver important blocks in given time and penalizes them if they did.</summary>
+        private async Task StallingLoopAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellationSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(StallingLoopIntervalMs, this.cancellationSource.Token).ConfigureAwait(false);
```

Comment:
AsyncLoop is a candidate?

Reply:
Potentially. Though I'd avoid it since current implementation is cleaner without it.